src/nvim/diff.c

```diff
@@ -1753,11 +1758,534 @@ void diff_clear(tabpage_T *tp)
   diff_T *next_p;
   for (p = tp->tp_first_diff; p != NULL; p = next_p) {
     next_p = p->df_next;
+    xfree(p->df_comparisonlines);
     xfree(p);
   }
   tp->tp_first_diff = NULL;
 }
 
+///
+/// return true if the options are set to use diff linematch
+///
+bool diff_linematch(diff_T *dp)
+{
+  if (!(diff_flags & DIFF_LINEMATCH)) {
+        return 0;
+  }
+  // are there more than three diff buffers?
+  int diffbuffers = 0;
+  int tsize = 0;
+  for (int i = 0; i < DB_COUNT; i++) {
+    if ( curtab->tp_diffbuf[i] != NULL ) {
+      diffbuffers++;
+      tsize += dp->df_count[i];
+    }
+  }
+  // avoid allocating a huge array because it will lag
+  if (tsize > linematch_lines) {
+    return 0;
+  }
+  if (diffbuffers <= 3) {  // can diff up to 3 buffers
+    return 1;
+  }
+  return 0;
+}
+
+///
+/// Count the number of virtual (filler + non filler) lines between "start" and
+/// "endline" in "win"
+///
+/// @param win
+/// @param start
+/// @param endline
+int count_virtual_lines(win_T *win, linenr_T start, linenr_T endline)
+{
+  int virtual_lines = 0;
+  for (int k = start; k <= endline; k++) {
+    int n = diff_check(win, k, NULL);
+    if (n > 0) {
+        virtual_lines+=n;  // filler lines
+    }
+    virtual_lines++;
+  }
+  return virtual_lines;
+}
+/// in "win" window, move from "lnum" down by the amount "virtual_lines"
+/// and return the number of real lines moved. if a non null pointer is
+/// passed to "line_new_virtualp", it will be set to the number of virtual
+/// lines moved
+///
+/// @param win
+/// @param lnum
+/// @param virtual_lines
+/// @param line_new_virtualp
+int count_virtual_to_real(win_T *win, const linenr_T lnum,
+                          const int virtual_lines, int *line_new_virtualp )
+{
+  int real_offset = 0;
+  int virtual_offset = 0;
+  while (1) {
+    int n = diff_check(win, lnum+real_offset, NULL);
+    virtual_offset++;
+    if (n > 0) {
+        virtual_offset+=n;  // filler lines
+    }
+    if (virtual_offset > virtual_lines) {
+        break;
+    }
+    real_offset++;
+  }
+  if ( line_new_virtualp != NULL ) {
+    (*line_new_virtualp) = virtual_offset;
+  }
+  return real_offset;
+}
+
+long count_matched_chars_f(char_u** stringps, int* fromValues, int n, int*** comparison_mem) {
+
+  int matched_chars = 0, pointerindex = 0, matched = 0;
+  for (int i = 0; i < n; i++) {
+    for (int j = i + 1; j < n; j++) {
+      if ( stringps[i] != NULL && stringps[j] != NULL ) {
+        int i1 = fromValues[i]; // index of where to get the buffer
+        int j1 = fromValues[j];
+        if (comparison_mem[pointerindex][i1][j1] == -1) {
+          comparison_mem[pointerindex][i1][j1] = count_matched_chars(
+              stringps[i], stringps[j]
+              );
```

Comment:
Line length limit is now 100
```suggestion
          comparison_mem[pointerindex][i1][j1] = count_matched_chars(stringps[i], stringps[j]);
```